### PR TITLE
Fix occasional buffering for http live streams

### PR DESCRIPTION
**Pull request by a1rwulf**
When watching live http streams often the stream starts (taking ~4sec) and after a few seconds of playback we run out of data and buffer again for around 5 seconds.
My usecase is mostly about playback through pvr.iptvsimple, when using this pvr addon the FileItem's Path is set to pvr:// while the actual item (=DynPath) is the http:// stream.
I let InputStreamFFmpeg handle those and set the Realtime flag which removes the buffering issue.

While looking at it I tried to do minor cleanup as well.
Additionally I found that the InputStreamFactory modifies the FileItem's path while this must not be done afaik.

## How Has This Been Tested?
Tested on linux x11:

- Playback of http streams using a m3u playlist -> no change to current behaviour.
- Playback of http streams using pvr.iptvsimple -> InputStreamFFmpeg is used and VP treats it as realtime stream
- Playback of nfs mkv -> no change

## Types of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp [hunk 1]
```diff
@@ -97,33 +95,32 @@ std::shared_ptr<CDVDInputStream> CDVDFactoryInputStream::CreateInputStream(IVide
...
+          StringUtils::StartsWithNoCase(file, "mmst://") ||
+          StringUtils::StartsWithNoCase(file, "mmsh://") ||
+          StringUtils::StartsWithNoCase(file, "rtmp://") ||
+          StringUtils::StartsWithNoCase(file, "rtmpt://") ||
+          StringUtils::StartsWithNoCase(file, "rtmpe://") ||
+          StringUtils::StartsWithNoCase(file, "rtmpte://") ||
+          StringUtils::StartsWithNoCase(file, "rtmps://"))
+  {
     return std::shared_ptr<CDVDInputStreamFFmpeg>(new CDVDInputStreamFFmpeg(fileitem));
-#ifdef ENABLE_DVDINPUTSTREAM_STACK
```
Comment (Rechi) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp]: behaviour change shouldn't be committed as `remove unnecessary ifdef`

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamFFmpeg.cpp [hunk 2]
```diff
@@ -155,3 +139,10 @@ std::string CDVDInputStreamFFmpeg::GetFileName()
   }
   return CDVDInputStream::GetFileName();
 }
+
+void CDVDInputStreamFFmpeg::SetRealtime(bool realtime)
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStreamFFmpeg.cpp]: the assumptions you take here are wrong i.e. realtime streams can be paused or seeked back if timeshift is active

File: xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp [hunk 3]
```diff
@@ -164,6 +161,13 @@ std::shared_ptr<CDVDInputStream> CDVDFactoryInputStream::CreateInputStream(IVide
 
     if (URIUtils::IsProtocol(finalFileitem.GetPath(), "udp"))
       return std::shared_ptr<CDVDInputStreamFFmpeg>(new CDVDInputStreamFFmpeg(finalFileitem));
+
+    if (StringUtils::StartsWithNoCase(fileitem.GetPath(), "pvr://") && URIUtils::IsProtocol(finalFileitem.GetDynPath(), "http"))
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp]: there are case where is assumtion is not correct.

**Comment by FernetMenta** [comment]
You should look for a way to set the realtime field for fileitems passed by iptvsimple addon. Maybe set a property on the fileitem

**Comment by peak3d** [comment]
Got the information that channel switch takes now 7 seconds for PVR.
So something is not right here

**Comment by M2k80** [comment]
i tried the first version of this PR, it fixes the 5 sek buffering on pvr.vuplus also, but open demuxer now takes 6 seconds.
**its only an info**

**Comment by a1rwulf** [comment]
That's why I marked it as RFC and WIP.
I'll rework this accordingly.

**Comment by ksooo** [comment]
@FernetMenta imo, setting isrealtime in getsreamproperties pvr api function won't work as this is called only once before the stream shall be opened. But realtime flag is a dynamic property of the stream. Value can change at any time when entering / leaving timeshift mode.

And isrealtimestream API function only gets called  on pvr input streams, right? But iptvsimple  streams  are not served using pvrinputstream.

Currently,  I have  no idea how to fix this properly.

**Comment by ksooo** [comment]
Wait, maybe it's as easy as always setting the "realtime" flag at the file item when getting a live stream URL from the add-on as result of a getlivestreamproperties API call? When open a pvr channel, it should initially always be "realtime", right?

**Comment by FernetMenta** [comment]
I won't tie it to PVR. A generic item property can also be set in a strm file.

**Comment by ksooo** [comment]
> A generic item property can also be set in a strm file.

 Yes, that's my idea.

> I won't tie it to PVR.

 It's tied to PVR in a way hat PVR component sets the item property. No special PVR hack in VP needed for that.

**Comment by ksooo** [comment]
I will come up with a PR - seems to be the best way to show you my idea.

**Comment by ksooo** [comment]
> I will come up with a PR - seems to be the best way to show you my idea.

Unfortunately, we will need a PVR addon API change to fix this properly. This is high effort as all PVR addons needs ned to be adjusted. I will take of this as soon as my time allows.

**Comment by liberty-developer** [comment]
IIRC pvr.zattoo ises the generic listItem properties for steering such things
IMO there is no API change needed

**Comment by ksooo** [comment]
> IMO there is no API change needed

 There is, we don't need a new API and can use GetChannelStreamProperties function, but we need to define a new property. 

https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h#L82

**Comment by a1rwulf** [comment]
Thx @ksooo for the superseding PR.
@FernetMenta what about the first 2 commits?
Shall I split the first one as per @Rechi's comments?
Is the second one valid?

**Comment by FernetMenta** [comment]
first 2 commits look ok, imo. please submit in a separate pr

**Pull request closed without merging**